pkg/services/ngalert/notifier/historian.go

```diff
@@ -0,0 +1,190 @@
+}
+
+type remoteLokiClient interface {
+	Ping(context.Context) error
+	Push(context.Context, []lokiclient.Stream) error
+}
+
+type NotificationHistorian struct {
+	client         remoteLokiClient
+	externalLabels map[string]string
+	metrics        *metrics.NotificationHistorian
+	log            log.Logger
+}
+
+func NewNotificationHistorian(logger log.Logger, cfg lokiclient.LokiConfig, req client.Requester, metrics *metrics.NotificationHistorian, tracer tracing.Tracer) *NotificationHistorian {
+	return &NotificationHistorian{
+		client:         lokiclient.NewLokiClient(cfg, req, metrics.BytesWritten, metrics.WriteDuration, logger, tracer, LokiClientSpanName),
+		externalLabels: cfg.ExternalLabels,
+		metrics:        metrics,
+		log:            logger,
+	}
+}
+
+func (h *NotificationHistorian) TestConnection(ctx context.Context) error {
+	return h.client.Ping(ctx)
+}
+
+func (h *NotificationHistorian) Record(ctx context.Context, alerts []*types.Alert, retry bool, notificationErr error, duration time.Duration) <-chan error {
+	stream, err := h.prepareStream(ctx, alerts, retry, notificationErr, duration)
+	logger := h.log.FromContext(ctx)
+	errCh := make(chan error, 1)
+	if err != nil {
+		logger.Error("Failed to convert notification history to stream", "error", err)
+		errCh <- fmt.Errorf("failed to convert notification history to stream: %w", err)
+		close(errCh)
+		return errCh
+	}
+
+	// This is a new background job, so let's create a new context for it.
+	// We want it to be isolated, i.e. we don't want grafana shutdowns to interrupt this work
+	// immediately but rather try to flush writes.
+	// This also prevents timeouts or other lingering objects (like transactions) from being
+	// incorrectly propagated here from other areas.
+	writeCtx := context.Background()
+	writeCtx, cancel := context.WithTimeout(writeCtx, NotificationHistoryWriteTimeout)
+	writeCtx = trace.ContextWithSpan(writeCtx, trace.SpanFromContext(ctx))
+
+	go func(ctx context.Context) {
+		defer cancel()
+		defer close(errCh)
+		logger := h.log.FromContext(ctx)
+		logger.Debug("Saving notification history")
+		h.metrics.WritesTotal.Inc()
+
+		if err := h.recordStream(ctx, stream, logger); err != nil {
+			logger.Error("Failed to save notification history", "error", err)
+			h.metrics.WritesFailed.Inc()
+			errCh <- fmt.Errorf("failed to save notification history: %w", err)
+		}
+	}(writeCtx)
+	return errCh
+}
+
+func (h *NotificationHistorian) prepareStream(ctx context.Context, alerts []*types.Alert, retry bool, notificationErr error, duration time.Duration) (lokiclient.Stream, error) {
+	receiverName, ok := notify.ReceiverName(ctx)
+	if !ok {
+		return lokiclient.Stream{}, fmt.Errorf("receiver name not found in context")
+	}
+	groupLabels, ok := notify.GroupLabels(ctx)
+	if !ok {
+		return lokiclient.Stream{}, fmt.Errorf("group labels not found in context")
+	}
+	now, ok := notify.Now(ctx)
+	if !ok {
+		return lokiclient.Stream{}, fmt.Errorf("now not found in context")
+	}
```

Comment:
I am not fan of this as this loosely couples us to the guts of the Alertmanager. Let's change the interface in the alerting module in the notifier and collect all require information there.